include/seqan3/alignment/matrix/detail/score_matrix_single_column.hpp

```diff
@@ -0,0 +1,246 @@
+ */
+template <typename score_t>
+class score_matrix_single_column<score_t>::matrix_iterator
+{
+private:
+
+    //!\brief The type of the zipped score column.
+    using matrix_column_t = decltype(views::zip(std::declval<physical_column_t &>(),
+                                                std::declval<physical_column_t &>(),
+                                                std::declval<virtual_column_t &>()));
+
+    //!\brief The transform adaptor to convert the tuple from the zip view into a seqan3::detail::affine_cell_type.
+    static constexpr auto transform_to_affine_cell = std::views::transform([] (auto && tpl)
+    {
+        using fwd_tuple_t = decltype(tpl);
+        return affine_cell_proxy<remove_cvref_t<fwd_tuple_t>>{std::forward<fwd_tuple_t>(tpl)};
+    });
+
+    //!\brief The pointer to the underlying matrix.
+    score_matrix_single_column * m_host_ptr{nullptr};
+    //!\brief The current column stored as a position.
+    size_t m_current_column{};
+    //!\brief The column index of the column behind the last one.
+    size_t m_end_column{};
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief The value type.
+    using value_type = decltype(std::declval<matrix_column_t>() | transform_to_affine_cell);
+    //!\brief The reference type.
+    using reference = value_type;
+    //!\brief The pointer type.
+    using pointer = void;
+    //!\brief The difference type.
+    using difference_type = std::ptrdiff_t;
+    //!\brief The iterator category.
+    using iterator_category = std::input_iterator_tag;
+    //!\}
+
+    /*!\name Constructor, assignment and destructor
+     * \{
+     */
+    matrix_iterator() noexcept = default; //!< Defaulted.
+    matrix_iterator(matrix_iterator const &) noexcept = default; //!< Defaulted.
+    matrix_iterator(matrix_iterator &&) noexcept = default; //!< Defaulted.
+    matrix_iterator & operator=(matrix_iterator const &) noexcept = default; //!< Defaulted.
+    matrix_iterator & operator=(matrix_iterator &&) noexcept = default; //!< Defaulted.
+    ~matrix_iterator() = default; //!< Defaulted.
+
+    /*!\brief Initialises the iterator from the underlying matrix.
+     *
+     * \param[in] host_matrix The underlying matrix.
+     */
+    explicit matrix_iterator(score_matrix_single_column & host_matrix) noexcept :
+        m_host_ptr{std::addressof(host_matrix)},
+        m_end_column{m_host_ptr->m_number_of_columns}
+    {}
+    //!\}
+
+    /*!\name Access operators
+     * \{
+     */
+    //!\brief Returns the range over the current column.
+    reference operator*() const
+    {
+        return views::zip(m_host_ptr->m_optimal_column, m_host_ptr->m_horizontal_column, m_host_ptr->m_vertical_column)
+             | transform_to_affine_cell;
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\brief Move `this` to the next column.
+    matrix_iterator & operator++()
+    {
+        ++m_current_column;
+        return *this;
+    }
```

Comment:
Wouldn't you need to also increment `m_host_ptr`?
Otherwise you always return the same column?

Reply:
That's the semantics of this implementation. It only fakes a real matrix but in reality it only stores a single column that will be used as the next column. So the algorithm is basically overwriting the old column, since we don't need access to it anymore. At least the algorithm knows about it and this is strongly coupled with the algorithm